File: ref_leak.py
class C:
    ref = None

    def __del__(self):
        print "__del__ " + str(self)


def test():
    a = C()
    b = C()

    a.ref = b
    b.ref = a

    print "{0}.ref = {1}".format(a, a.ref)
    print "{0}.ref = {1}".format(b, b.ref)


test()

File: ref_leak.pl
use v5.14;

package C {
    sub new {
        my ($class, $ref) = @_;
        bless { ref => undef } => $class;
    }

    sub ref {
        my ($self, $ref) = @_;
        $self->{ref} = $ref;
    }

    sub DESTROY {
        warn "DESTROY @_";
    }
}


sub test {
    my $a = C->new;
    my $b = C->new;

    $a->ref($b);
    $b->ref($a);

    warn "$a\->ref = $a->{ref}";
    warn "$b\->ref = $b->{ref}";
}

test();
say "END";
